range.c

```diff
@@ -361,42 +390,53 @@ range_step_size(VALUE range, VALUE args, VALUE eobj)
 }
 
 /*
- *  Document-method: Range#step
- *  Document-method: Range#%
  *  call-seq:
- *     rng.step(n=1) {| obj | block }    -> rng
- *     rng.step(n=1)                     -> an_enumerator
- *     rng.step(n=1)                     -> an_arithmetic_sequence
- *     rng % n                           -> an_enumerator
- *     rng % n                           -> an_arithmetic_sequence
+ *    step(n = 1) {|element| ... } -> self
+ *    step(n = 1)                  -> enumerator
  *
- *  Iterates over the range, passing each <code>n</code>th element to the block.
- *  If begin and end are numeric, +n+ is added for each iteration.
- *  Otherwise #step invokes #succ to iterate through range elements.
+ *  Iterates over the elements of +self+.
  *
- *  If no block is given, an enumerator is returned instead.
- *  Especially, the enumerator is an Enumerator::ArithmeticSequence
- *  if begin and end of the range are numeric.
+ *  <b>For a \Numeric \Range</b>
+ *
+ *  With a block given and no argument,
+ *  calls the block each element of the range; returns +self+:
+ *
+ *    a = []
+ *    (1..5).step {|element| a.push(element) } # => 1..5
+ *    a # => [1, 2, 3, 4, 5]
+ *
+ *  With a block given and a positive integer argument +n+ given,
+ *  calls the block with element +0+, element +n+, element <tt>2n</tt>, and so on:
+ *
+ *    a = []
+ *    (1..5).step(2) {|element| a.push(element) } # => 1..5
+ *    a # => [1, 3, 5]
+ *
+ *  With no block given, returns a specialized enumerator:
+ *
+ *    e = (1..5).step(2) # => ((1..5).step(2))
+ *    e.class            # => Enumerator::ArithmeticSequence
+ *
+ *  <b>For a Non-Numeric \Range</b>
```

Comment:
It looks like the behavior for the numeric range and non-numeric range case is almost identical, except that numeric ranges will return `Enumerator::ArithmeticSequence` for `step` calls without blocks (non-numeric ranges will return plain `Enumerator`).  The documentation should reflect that, and only document cases differently if the behavior is the differs between the cases.

Reply:
Reworked.